sled-agent/src/services.rs

```diff
@@ -773,6 +773,100 @@ pub struct ServiceManager {
     inner: Arc<ServiceManagerInner>,
 }
 
+/// Ensure that a NAT entry exists, overwriting a previous conflicting entry if
+/// applicable.
+///
+/// nat_ipv\[46\]_create are not idempotent (see oxidecomputer/dendrite#343),
+/// but this wrapper function is. Call this from sagas instead.
+#[allow(clippy::too_many_arguments)]
+async fn dpd_ensure_nat_entry(
+    client: &DpdClient,
+    log: &Logger,
+    target_ip: IpAddr,
+    target_mac: DpdTypes::MacAddr,
+    target_first_port: u16,
+    target_last_port: u16,
+    target_vni: u32,
+    sled_ip_address: &std::net::Ipv6Addr,
+) -> Result<(), Error> {
+    let existing_nat = match &target_ip {
+        IpAddr::V4(ip) => client.nat_ipv4_get(ip, target_first_port).await,
+        IpAddr::V6(ip) => client.nat_ipv6_get(ip, target_first_port).await,
+    };
+
+    // If a NAT entry already exists, but has the wrong internal
+    // IP address, delete the old entry before continuing (the
+    // DPD entry-creation API won't replace an existing entry).
+    // If the entry exists and has the right internal IP, there's
+    // no more work to do for this external IP.
+    match existing_nat {
+        Ok(existing) => {
+            let existing = existing.into_inner();
+            if existing.internal_ip != *sled_ip_address {
+                info!(log, "deleting old nat entry";
+                      "target_ip" => ?target_ip);
+
+                match &target_ip {
+                    IpAddr::V4(ip) => {
+                        client.nat_ipv4_delete(ip, target_first_port).await
+                    }
+                    IpAddr::V6(ip) => {
+                        client.nat_ipv6_delete(ip, target_first_port).await
+                    }
+                }?;
+            } else {
+                info!(log,
+                      "nat entry with expected internal ip exists";
+                      "target_ip" => ?target_ip,
+                      "existing_entry" => ?existing);
+
+                return Ok(());
+            }
+        }
+        Err(e) => {
+            if e.status() == Some(http::StatusCode::NOT_FOUND) {
+                info!(log, "no nat entry found for: {target_ip:#?}");
```

Comment:
Should this be a `warn!` or `error!`?

Reply:
It doesn't look like it.  This routine just wants to ensure that  NAT entry exists, but doesn't have an opinion on whether it should already exist or not.  This could indicate an error, or it could just be that the entry was just added by nexus and this is our first attempt to push it to the switch.